### PR TITLE
csi: add custom ceph.conf for csi pods

### LINKED ISSUE #9220: Custom ceph.conf for cephcsi pods
Currently, the admin/user can configure the `ceph.conf` for daemon pods using https://rook.io/docs/rook/v1.7/ceph-advanced-configuration.html#custom-cephconf-settings. This issue is to have a custom `ceph.conf` for the clients/cephcsi pods. the support for the same is added in cephcsi https://github.com/ceph/ceph-csi/pull/2476. 

Below are the points I discussed with @travisn 
* Add new `csi-ceph-conf.yaml` to the example directory (this is a new configmap as this ceph.conf is applicable for only clients)
* If the configmap is created update cephcsi templates to mount the configmap to csi pods
* If the configmap is not created cephcsi will use the default `ceph.conf` what it creates
*  Update documentation for the new configmap

**Note:-** Even though the cephcsi/Rook operator can handle multiple ceph clusters. As cephcsi is installed only once per rook operator we can have only once `ceph.conf` for csi pods for all the ceph clusters.

Below is the little clarity am looking for. ( didn't get a chance to think about it before discussing it with Travis)

* Do we need to add owner-ref for this configmap or user will take care of deleting it manually as part of the cleanup process?

* How to handle the upgrade process. if cephcsi makes something detailed we just need to document the new changes so the user can update the `ceph.conf` he created?

@travisn please add if I missed anything from the discussion

cc @travisn @leseb @BlaineEXE

### PR DESCRIPTION
**Pull request by Madhu-1**
**Description of your changes:**
Currently, the admin/user can configure the ceph.conf for daemon pods using https://rook.io/docs/rook/v1.7/ceph-advanced-configuration.html#custom-cephconf-settings. This the above custom ceph.conf is only for ceph pods. the support to provide custom ceph.conf for cephcsi is added in cephcsi ceph/ceph-csi#2476

This PR adds the support to create/update ceph.conf for csi pods.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

**Which issue is resolved by this Pull Request:**
Resolves #9220

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

@travisn what is the best place to add this documentation? can I add this as a sub section under https://rook.io/docs/rook/v1.7/ceph-advanced-configuration.html#custom-cephconf-settings

### CODE REVIEW

File: pkg/operator/ceph/csi/spec.go [hunk 1]
```diff
@@ -336,6 +337,10 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
 		}
 	}
 
+	if AllowUnsupported || v.SupportsCustomCephConf() {
+		tp.MountCustomCephConf = true
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:341]:
```suggestion
		tp.MountCustomCephConf = v.SupportsCustomCephConf()
```

Looks like `AllowUnsupported` is tested in `SupportsCustomCephConf` already.

File: pkg/operator/ceph/csi/version.go [hunk 2]
```diff
@@ -112,3 +116,29 @@ func extractCephCSIVersion(src string) (*CephCSIVersion, error) {
...
 }
+
+// SupportsCustomCephConf checks if the detected version supports custom ceph.conf
+func (v *CephCSIVersion) SupportsCustomCephConf() bool {
+	// if AllowUnsupported is set also a csi-image greater than the supported ones are allowed
+	if AllowUnsupported {
+		return true
+	}
+
+	if !v.isAtLeast(&minimum) {
```
Comment (leseb) [inline pkg/operator/ceph/csi/version.go:127]: What's `minimum` here? Can the name be more explicit?

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/version.go (hunk 2 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/version.go]: it should be cephConfSupportedVersion fixed now.

**Comment by Madhu-1** [comment]
> @leseb any idea about commitlint failure?

fixed now its due to https://github.com/conventional-changelog/commitlint/issues/896

**Comment by BlaineEXE** [comment]
What about the possiblity of configuring the CSI clients in the mon config store so that different Ceph clusters can have different CSI overrides per [this conversation](https://github.com/rook/rook/issues/9220#issuecomment-983527667)?

File: pkg/operator/ceph/csi/ceph_config.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+func CreateCsiCephConfigMap(ctx context.Context, clientset kubernetes.Interface, ownerInfo *k8sutil.OwnerInfo, namespace string) error {
+	configMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configName,
+			Namespace: namespace,
+		},
+	}
+	configMap.Data = map[string]string{
+		cephCongKey: cephConfVal,
+		keyringKey:  keyringVal,
```
Comment (travisn) [inline pkg/operator/ceph/csi/ceph_config.go:65]: Why add an empty keyring to the configmap? Do you expect users to update with the real keyring? Or why is this needed? Seems like it should be in a secret if anywhere.

File: pkg/operator/ceph/csi/ceph_config.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+`
+	// keyring should be adds to the same configmap as ceph.conf as we cannot
+	// create keyring file in the same directory as ceph.conf.
+	keyringKey = "keyring"
+	// empty keyring file content
+	keyringVal = ""
+)
+
+// CreateCsiConfigMap creates an ceph.conf for ceph-csi. If a config map already
+// exists, it will return it.
```
Comment (travisn) [inline pkg/operator/ceph/csi/ceph_config.go:55]: So if someone wants a custom ceph.conf in the csi driver they have to create it first, or update the default after it is created here?

Comment (travisn) [reply pkg/operator/ceph/csi/ceph_config.go:55]: What if the default needs to change in the future? It will be difficult to know the difference between a previous default and user-customized contents. You might consider for the csi driver to merge its required settings with the user-specified settings, and keep this configmap empty by default for user settings.

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml [hunk 5]
```diff
@@ -170,6 +170,10 @@ spec:
               mountPath: /etc/ceph-csi-config/
             - name: keys-tmp-dir
               mountPath: /tmp/csi/keys
+            {{ if .MountCustomCephConf }}
```
Comment (travisn) [inline pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:173]: Would the rbd or cephfs plugins or provisioners ever need different ceph.conf?

**Review by travisn — CHANGES_REQUESTED** [review]
> @travisn what is the best place to add this documentation? can I add this as a sub section under https://rook.io/docs/rook/v1.7/ceph-advanced-configuration.html#custom-cephconf-settings

Yes sounds good to add another section below that one for the custom ceph.conf settings for the csi driver. It might be good as its own section instead of a subsection, but you can decide.

File: pkg/operator/ceph/csi/ceph_config.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	configName = "csi-ceph-conf-override"
+
+	cephCongKey = "ceph.conf"
```
Comment (leseb) [inline pkg/operator/ceph/csi/ceph_config.go:33]: the client package has a `DefaultConfigFile` we can use.

File: pkg/operator/ceph/csi/version.go [hunk 7]
```diff
@@ -112,3 +116,29 @@ func extractCephCSIVersion(src string) (*CephCSIVersion, error) {
 
 	return &CephCSIVersion{major, minor, bugfix}, nil
 }
+
+// SupportsCustomCephConf checks if the detected version supports custom ceph.conf
+func (v *CephCSIVersion) SupportsCustomCephConf() bool {
+	// if AllowUnsupported is set also a csi-image greater than the supported ones are allowed
+	if AllowUnsupported {
```
Comment (leseb) [inline pkg/operator/ceph/csi/version.go:123]: Why do we need this check? Can we wait for this code to merge after we use ceph-csi 3.5.0? So we don't need this?

File: pkg/operator/ceph/csi/ceph_config.go (hunk 3 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:65]: we don't use this keyring for csi operations, but if this file is not created we see a warning when we execute ceph CLI commands. that's the reason for having an empty file.

File: pkg/operator/ceph/csi/ceph_config.go (hunk 4 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:55]: >So if someone wants a custom ceph.conf in the csi driver they have to create it first, or update the default after it is created here?

both works if the configmap is in the expected format.

>What if the default needs to change in the future? It will be difficult to know the difference between a previous default and user-customized contents. You might consider for the csi driver to merge its required settings with the user-specified settings, and keep this configmap empty by default for user settings.

merging might be difficult one but we can document the new configuration that needs to be added as part of the upgrade process?

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml (hunk 5 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:173]: we haven't come across any such cases. till now we are using the same ceph.conf for both rbd and cephfs in cephcsi also

File: pkg/operator/ceph/csi/version.go (hunk 7 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/version.go:123]: This is to support the custom cephcsi who are building cephcsi with a different versions (ex:- OCS).

**Comment by Madhu-1** [comment]
> What about the possiblity of configuring the CSI clients in the mon config store so that different Ceph clusters can have different CSI overrides per [this conversation](https://github.com/rook/rook/issues/9220#issuecomment-983527667)?

@BlaineEXE  that will be the future enhancements. for now, we are going with a single ceph.conf for all ceph clusters.

**Comment by Madhu-1** [comment]
> > @travisn what is the best place to add this documentation? can I add this as a sub section under https://rook.io/docs/rook/v1.7/ceph-advanced-configuration.html#custom-cephconf-settings
> 
> Yes sounds good to add another section below that one for the custom ceph.conf settings for the csi driver. It might be good as its own section instead of a subsection, but you can decide.

Thanks, @travisn added as a new section.

**Review by leseb — COMMENTED** [review]
Can you add a test in `pkg/operator/ceph/csi/controller_test.go` that the CM has been created?

File: pkg/operator/ceph/csi/ceph_config.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+			Name:      configName,
+			Namespace: namespace,
+		},
+	}
+	configMap.Data = map[string]string{
+		client.DefaultConfigFile: cephConfVal,
+		keyringKey:               keyringVal,
+	}
+
+	err := ownerInfo.SetControllerReference(configMap)
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:68]: Yes, rook-operator deployment is the owner. yes once the rook operator is removed all CSI resources get deleted along with this configmap also.

[the remark this answers is not in the document]
Comment (leseb) [reply pkg/operator/ceph/csi/ceph_config.go:68]: Yes but we don't want to throw user changes away...

[the remark this answers is not in the document]
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:68]: if we don't set owner ref this will be left when the Rook ceph cluster is teardowned. won't it be a cleanup problem?

[the remark this answers is not in the document]
Comment (leseb) [reply pkg/operator/ceph/csi/ceph_config.go:68]: When the ceph cluster is teardowned the CM won't be deleted since it is not the owner, did you mean complete teardown operator included? I guess it depends, if the ns goes away it's fine, if the just cephcluster/operator goes away then the CM will remain...

[the remark this answers is not in the document]
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:68]: Yes, when the rook-operator deployment is deleted this should be deleted isn't it? or else the CM even with default configurations will be left in the namespace even if it's created by the rook operator.  I thought leaving stale resources is the problem for cleanup

File: pkg/operator/ceph/csi/ceph_config_test.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestCreateCsiCephConfigMap(t *testing.T) {
+	namespace := "test-namespace"
+
+	ctx := context.TODO()
+	// create mocked client
+	clientset := test.New(t, 3)
+
```
Comment (leseb) [inline pkg/operator/ceph/csi/ceph_config_test.go:37]: remove new line

File: pkg/operator/ceph/csi/ceph_config.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (leseb) [reply pkg/operator/ceph/csi/ceph_config.go:68]: I'd argue it's probably better to leave the CM than trash the user's settings.

[the remark this answers is not in the document]
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go]: I don't have any other option :+1: removed owner reference.

File: pkg/operator/ceph/csi/ceph_config.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/csi/ceph_config.go:54]: I don't think we should mix user settings and required settings in the same configmap or else we really can't control what happens during upgrade. We have run into this with the OCS operator and the existing ceph.conf overrides. If they want ceph.conf overrides for the csi driver, we need to reserve this config purely for the user config. Are the ceph client defaults not sufficient if we leave this blank?

File: pkg/operator/ceph/csi/ceph_config.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	configName = "csi-ceph-conf-override"
+
+	cephConfVal = `[global]
```
Comment (travisn) [inline pkg/operator/ceph/csi/ceph_config.go:33]: Are these default settings coming from a downstream requirement? We've never had default ceph.conf settings before, why do we need them now?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/ceph_config.go (hunk 10 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:33]: These are the defaults used in cephcsi from long time.

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml [hunk 11]
```diff
@@ -130,6 +130,10 @@ spec:
               mountPath: /etc/ceph-csi-config/
             - name: keys-tmp-dir
               mountPath: /tmp/csi/keys
+            {{ if .MountCustomCephConf }}
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml:133]: No it wiill fail, because cephcsi will try to write to the mount point and write will fail

File: pkg/operator/ceph/csi/ceph_config.go (hunk 4 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:54]: @travisn any idea about this one? it's difficult to merge I think. or can we introduce ceph.conf configmap per ceph cluster to override defaults? if user creates a configmap in the ceph cluster namespace with the default name we will get and replace the content of ceph.conf we are creating.  any better idea how to solve this one?

File: Documentation/ceph-advanced-configuration.md [hunk 12]
```diff
@@ -293,6 +294,87 @@ data:
...
+
+After the CSI pods are restarted, the new settings should be in effect.
+
+### Example
+
+In this example we will set the `rbd_validate_pool` to `false` to skip rbd pool
+validation
+
+> **WARNING**: Modify Ceph settings carefully avoid modifying the default
+> configuration. You are leaving the sandbox tested by Rook.
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply Documentation/ceph-advanced-configuration.md:321]: copy paste from Rook section removed it.

File: Documentation/ceph-advanced-configuration.md [hunk 13]
```diff
@@ -293,6 +294,87 @@ data:
...
+> Changing the settings could result in unexpected results or even data loss if used incorrectly.
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply Documentation/ceph-advanced-configuration.md:322]: removed copy paste mistake.

File: Documentation/ceph-advanced-configuration.md [hunk 14]
```diff
@@ -293,6 +294,87 @@ data:
...
+    auth_client_required = cephx
+
+    fuse_set_user_groups = false
+    fuse_big_writes = true
+  # keyring is a required key and its value should be empty
+  keyring: |
+```
+
+To apply your desired configuration, you will need to update this ConfigMap. The next time the
+CephCSI pod(s) start, they will use the updated configs.
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply Documentation/ceph-advanced-configuration.md:348]: Yes, correct. even it gets auto-updated. asking to wait for restart to make sure it's updated as ceph.conf will get applied if remounts happen.

**Review by leseb — APPROVED** [review]

File: Documentation/ceph-advanced-configuration.md [hunk 15]
```diff
@@ -293,6 +294,83 @@ data:
...
+```console
+kubectl -n rook-ceph edit configmap csi-ceph-conf-override
+```
+
+> **WARNING**: Don't delete/modify the keyring value it should be left empty.
+> Changing keyring settings could result in unexpected results.
+
+Modify the settings and save. Each line you add should be indented from the `ceph.conf` property as such:
+
+```yaml
```
Comment (travisn) [inline Documentation/ceph-advanced-configuration.md:355]: Can we keep this section simpler by having just the one example above? Showing a single line added could be done in the example above.

File: pkg/operator/ceph/csi/ceph_config.go (hunk 10 above)
Comment (travisn) [reply pkg/operator/ceph/csi/ceph_config.go:33]: By default, what if we create an empty configmap in rook, and if it's empty then the csi driver sets its default?

File: pkg/operator/ceph/csi/ceph_config.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/csi/ceph_config.go:54]: What about the following?
- Create an empty configmap by default? 
- If the CSI driver finds an empty configmap, set to its default
- If the CSI driver finds settings in the configmap, use it (and don't set the default)

This way the default stays with the csi driver, and the user can override to anything and we don't need to worry about merging. But the main advantage I see is that the default can be updated. This PR doesn't allow the default to ever be updated since we won't reconcile the configmap after creation.

File: pkg/operator/ceph/csi/ceph_config.go (hunk 10 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:33]: not possible as the csi driver won't interact with KUbe api's Rook is mounting the configmap as a file to the csi pods. even if the file is empty csi driver cannot write due to a permission issue.

File: pkg/operator/ceph/csi/ceph_config.go (hunk 4 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/ceph_config.go:54]: due to mounting configmap as a file cephcsi cannot write back to the configmap? let me do double check if its possible

Comment (travisn) [reply pkg/operator/ceph/csi/ceph_config.go:54]: The CSI driver would need to mount this configmap to some location other than `/etc/ceph/ceph.conf` so that the csi driver could either copy from the other location to here, or the csi driver would write its default just like it was doing previously.

**Comment by Madhu-1** [comment]
>By default, what if we create an empty configmap in rook, and if it's empty then the csi driver sets its default?

configmap mounts are always read-only in kubernetes. CSI Driver cannot do write operation to the configmap mounted files, i have seen permission problem

```bash
sh-4.4# echo tesing> /etc/ceph/ceph.conf
sh: /etc/ceph/ceph.conf: Read-only file system
sh-4.4# ls -lrt /etc/ceph/ceph.conf
lrwxrwxrwx 1 root root 16 Feb 15 05:45 /etc/ceph/ceph.conf -> ..data/ceph.conf
sh-4.4# 
```

**Comment by Madhu-1** [comment]
>What about the following?
>* Create an empty configmap by default?
>* If the CSI driver finds an empty configmap, set to its default
>* If the CSI driver finds settings in the configmap, use it (and don't set the default)
>* This way the default stays with the csi driver, and the user can override to anything and we don't need to worry about merging. But the main advantage I see is that the default can be updated. This PR doesn't allow the default to ever be updated since we won't reconcile the configmap after creation.

I can think of a way to solve this one.

* Create empty configmap by Rook
* When starting the csidriver check configmap is empty or not, if it's empty don't mount the configmap, if it's not empty mount the configmap
* As the configmap is not mounted the csidriver will set its defaults and use it and if a user updates the configmap he needs to restart the Rook operator pod to make sure the new csidriver deployment is created.

**Comment by travisn** [comment]
Per our discussion, that plan sounds good with the modification that we just check if the configmap exists instead of checking if it is empty. Then we don't need to worry about creating an empty configmap that might never be used.

**Comment by leseb** [comment]
> Per our discussion, that plan sounds good with the modification that we just check if the configmap exists instead of checking if it is empty. Then we don't need to worry about creating an empty configmap that might never be used.

Yet to be implemented?

**Comment by Madhu-1** [comment]
> > Per our discussion, that plan sounds good with the modification that we just check if the configmap exists instead of checking if it is empty. Then we don't need to worry about creating an empty configmap that might never be used.
> 
> Yet to be implemented?

Yes yet to be added, will make the changes and push changes by EOD.

File: Documentation/ceph-advanced-configuration.md [hunk 16]
```diff
@@ -293,6 +294,56 @@ data:
...
+add the settings to the ConfigMap after the cluster has been initialized,
+you can restart the Rook operator pod and wait for Rook to recreate CSI pods
+to take immediate effect.
+
+After the CSI pods are restarted, the new settings should be in effect.
+
+### Example
+
+In this [Example](https://github.com/rook/rook/tree/{{ branchName }}/deploy/csi-ceph-conf-override.yaml) we
+will set the `rbd_validate_pool` to `false` to skip rbd pool validation
```
Comment (travisn) [inline Documentation/ceph-advanced-configuration.md:315]:
```suggestion
will set the `rbd_validate_pool` to `false` to skip rbd pool validation.
```

File: Documentation/ceph-advanced-configuration.md [hunk 17]
```diff
@@ -293,6 +294,56 @@ data:
...
+  name: csi-ceph-conf-override
+  namespace: rook-ceph
+data:
+  ceph.conf: |
+    [global]
+    auth_cluster_required = cephx
+    auth_service_required = cephx
+    auth_client_required = cephx
+    rbd_validate_pool = false
+  # keyring is a required key and its value should be empty
```
Comment (travisn) [inline Documentation/ceph-advanced-configuration.md:335]: Why should the keyring be empty?

File: Documentation/ceph-advanced-configuration.md [hunk 18]
```diff
@@ -293,6 +294,56 @@ data:
...
+  keyring: |
+```
+
+> **WARNING**: Don't delete/modify the keyring value it should be left empty.
+> Changing keyring settings could result in unexpected results.
+
+```console
+kubectl create -f csi-ceph-conf-override.yaml
+```
+Restart Rook operator pod and wait for CSI pods to recreate.
```
Comment (travisn) [inline Documentation/ceph-advanced-configuration.md:345]:
```suggestion
Restart the Rook operator pod and wait for the CSI pods to be recreated.
```

File: Documentation/ceph-advanced-configuration.md [hunk 19]
```diff
@@ -293,6 +294,56 @@ data:
...
+
+In this [Example](https://github.com/rook/rook/tree/{{ branchName }}/deploy/csi-ceph-conf-override.yaml) we
+will set the `rbd_validate_pool` to `false` to skip rbd pool validation
+
+> **WARNING**: Modify Ceph settings carefully to avoid modifying the default
+> configuration.
+> Changing the settings could result in unexpected results if used incorrectly.
+
+```console
+cat csi-ceph-conf-override.yaml
```
Comment (travisn) [inline Documentation/ceph-advanced-configuration.md:322]: I might suggest just add a link to the example yaml and not duplicate it here. But if you want to keep it here, no need for the `cat` command
```suggestion
```

File: pkg/operator/ceph/csi/version.go [hunk 20]
```diff
@@ -112,3 +116,29 @@ func extractCephCSIVersion(src string) (*CephCSIVersion, error) {
...
+		return true
+	}
+
+	if !v.isAtLeast(&cephConfSupportedVersion) {
+		return false
+	}
+
+	if v.Major > cephConfSupportedVersion.Major {
```
Comment (travisn) [inline pkg/operator/ceph/csi/version.go:131]: Doesn't `v.isAtLeast()` on line 127 already cover lines 131-141?

File: pkg/operator/ceph/csi/controller.go [hunk 21]
```diff
@@ -194,6 +194,13 @@ func (r *ReconcileCSI) reconcile(request reconcile.Request) (reconcile.Result, e
 		return opcontroller.ImmediateRetryResult, errors.Wrap(err, "failed to create pool ID mapping config map")
 	}
 
+	exists, err := checkCsiCephConfigMapExists(r.opManagerContext, r.context.Clientset, r.opConfig.OperatorNamespace)
```
Comment (travisn) [inline pkg/operator/ceph/csi/controller.go:197]: How about if we create the configmap in at least one of the integration test suites? Then we can get some test coverage that the example works end to end both with the custom overrides and without.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/controller.go [hunk 22]
```diff
@@ -194,6 +194,13 @@ func (r *ReconcileCSI) reconcile(request reconcile.Request) (reconcile.Result, e
...
+	if err != nil {
+		return opcontroller.ImmediateRetryResult, errors.Wrap(err, "failed to get csi ceph.conf configmap")
+	}
+	if exists {
+		CustomCSICephConfigExists = true
```
Comment (leseb) [inline pkg/operator/ceph/csi/controller.go:202]: How is the configmap content being configured?

File: Documentation/ceph-advanced-configuration.md (hunk 17 above)
Comment (Madhu-1) [reply Documentation/ceph-advanced-configuration.md:335]: removed it now. Mounting configmap as subpath to csi pods.

File: Documentation/ceph-advanced-configuration.md (hunk 19 above)
Comment (Madhu-1) [reply Documentation/ceph-advanced-configuration.md:322]: Removed the cat and content

File: pkg/operator/ceph/csi/controller.go (hunk 22 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go:202]: The user needs to set the configuration in the csi configmap, if the configmap is present Rook will mount the file to the cephcsi container to `/etc/ceph/ceph.conf`.

File: pkg/operator/ceph/csi/version.go (hunk 20 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/version.go:131]: Yes this seems like duplicate, Removed it now.

File: pkg/operator/ceph/csi/controller.go (hunk 21 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go:197]: @travisn where exactly I can create this configmap in the integration test suite?

File: Documentation/ceph-advanced-configuration.md (hunk 16 above)
Comment (travisn) [reply Documentation/ceph-advanced-configuration.md:315]: nit

File: pkg/operator/ceph/csi/controller.go (hunk 21 above)
Comment (travisn) [reply pkg/operator/ceph/csi/controller.go:197]: How about in [CreateCephCluster()](https://github.com/rook/rook/blob/9840273d8e82ab5077f298aa7a6af11a801c4299/tests/framework/installer/ceph_installer.go#L238) next to where the configmap is created for the other ceph.conf overrides.

Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go:197]: Done. Thanks for the hint

**Comment by Madhu-1** [comment]
@travisn @leseb changes are done PTAL.

File: Documentation/ceph-advanced-configuration.md [hunk 23]
```diff
@@ -293,6 +294,36 @@ data:
     osd pool default size = 2
 ```
 
+## Custom CSI ceph.conf Settings
+
+> **WARNING**: It is highly recommended to use the default setting that comes with
+> CephCSI and this can only be used when absolutely necessary.
+> The `ceph.conf` be reset back to default values if/when the configurations are no
```
Comment (travisn) [inline Documentation/ceph-advanced-configuration.md:301]:
```suggestion
> The `ceph.conf` should be reset back to default values if/when the configurations are no
```

File: pkg/operator/ceph/csi/controller.go (hunk 22 above)
Comment (travisn) [reply pkg/operator/ceph/csi/controller.go:202]: How about always setting the value, even if it doesn't exist?
```suggestion
		CustomCSICephConfigExists = exists
```

File: tests/framework/installer/ceph_installer.go [hunk 24]
```diff
@@ -258,6 +258,21 @@ func (h *CephInstaller) CreateCephCluster() error {
 		return errors.Errorf("failed to create custom ceph.conf. %+v", err)
 	}
 
+	customCSISettings := map[string]string{
+		"ceph.conf": "[global]\nauth_client_required = cephx",
```
Comment (travisn) [inline tests/framework/installer/ceph_installer.go:262]: Does ceph parse the `\n`? How about this?
```suggestion
		"ceph.conf": `[global]
auth_client_required = cephx`,
```

**Review by travisn — CHANGES_REQUESTED** [review]
some minor suggestions

File: tests/framework/installer/ceph_installer.go (hunk 24 above)
Comment (Madhu-1) [reply tests/framework/installer/ceph_installer.go:262]: configmap will get created as required. I have tested this manually and even E2E is also working as this config is mounted to csi pods and it's also used for PVC creation etc.

**Review by travisn — COMMENTED** [review]
LGTM. Is this needed to be backported? Or shall we keep it for the 1.9 release? If not backporting, how about adding this to the pending release notes?

**Comment by Madhu-1** [comment]
> LGTM. Is this needed to be backported? Or shall we keep it for the 1.9 release? If not backporting, how about adding this to the pending release notes?

should be fine for 1.9 release added to pending notes.
